From 19769a917382da7d83f794ea2954eb956fd6130d Mon Sep 17 00:00:00 2001
From: Nikolay Vakhlyarskiy <nvakhlyarskiy@gmail.com>
Date: Fri, 3 May 2019 14:25:13 -0700
Subject: [PATCH 0630/1851] v4l: xilinx: dma: add v4l2_ctrl_handler

commit b42ef098aca9a0b95baf43ff10193077db10fd1b from
https://github.com/Xilinx/linux-xlnx.git

According to:

https://linuxtv.org/downloads/v4l-dvb-apis/kapi/v4l2-controls.html

1.12.4. Inheriting Controls
When a sub-device is registered with a V4L2 driver by calling
v4l2_device_register_subdev() and the ctrl_handler fields of both
v4l2_subdev and v4l2_device are set, then the controls of the
subdev will become automatically available in the V4L2 driver
as well. If the subdev driver contains controls that already exist
in the V4L2 driver, then those will be skipped (so a V4L2 driver
can always override a subdev control).

What happens here is that v4l2_device_register_subdev()
calls v4l2_ctrl_add_handler() adding the controls of the
subdev to the controls of v4l2_device.

Signed-off-by: Nikolay Vakhlyarskiy <nvakhlyarskiy@gmail.com>
Reviewed-by: Hyun Kwon <hyun.kwon@xilinx.com>
State: pending
Signed-off-by: Yaliang Wang <Yaliang.Wang@windriver.com>
---
 drivers/media/platform/xilinx/xilinx-dma.c | 7 +++++++
 drivers/media/platform/xilinx/xilinx-dma.h | 4 ++++
 2 files changed, 11 insertions(+)

diff --git a/drivers/media/platform/xilinx/xilinx-dma.c b/drivers/media/platform/xilinx/xilinx-dma.c
index 95ef373d9e1e..bb783d953224 100644
--- a/drivers/media/platform/xilinx/xilinx-dma.c
+++ b/drivers/media/platform/xilinx/xilinx-dma.c
@@ -1253,6 +1253,13 @@ int xvip_dma_init(struct xvip_composite_device *xdev, struct xvip_dma *dma,
 	if (ret < 0)
 		goto error;
 
+	ret = v4l2_ctrl_handler_init(&dma->ctrl_handler, 0);
+	if (ret < 0) {
+		dev_err(dma->xdev->dev, "failed to initialize V4L2 ctrl\n");
+		goto error;
+	}
+	dma->video.ctrl_handler = &dma->ctrl_handler;
+
 	/* ... and the video node... */
 	dma->video.fops = &xvip_dma_fops;
 	dma->video.v4l2_dev = &xdev->v4l2_dev;
diff --git a/drivers/media/platform/xilinx/xilinx-dma.h b/drivers/media/platform/xilinx/xilinx-dma.h
index 61c26ab103f7..4a5966c4d943 100644
--- a/drivers/media/platform/xilinx/xilinx-dma.h
+++ b/drivers/media/platform/xilinx/xilinx-dma.h
@@ -18,6 +18,7 @@
 #include <linux/videodev2.h>
 
 #include <media/media-entity.h>
+#include <media/v4l2-ctrls.h>
 #include <media/v4l2-dev.h>
 #include <media/videobuf2-v4l2.h>
 
@@ -56,6 +57,7 @@ static inline struct xvip_pipeline *to_xvip_pipeline(struct media_entity *e)
  * @video: V4L2 video device associated with the DMA channel
  * @pad: media pad for the video device entity
  * @remote_subdev_med_bus: media bus format of sub-device
+ * @ctrl_handler: V4L2 ctrl_handler for inheritance ctrls from subdev
  * @xdev: composite device the DMA channel belongs to
  * @pipe: pipeline belonging to the DMA channel
  * @port: composite device DT node port number for the DMA channel
@@ -80,6 +82,8 @@ struct xvip_dma {
 	struct media_pad pad;
 	u32 remote_subdev_med_bus;
 
+	struct v4l2_ctrl_handler ctrl_handler;
+
 	struct xvip_composite_device *xdev;
 	struct xvip_pipeline pipe;
 	unsigned int port;
-- 
2.31.1

